### PR TITLE
fix(tokens): reconcile transcript usage per message.id (2-4x inflated token totals and costs)

### LINKED ISSUE #293: Token totals and costs inflated 2-4x: usage summed per JSONL record instead of reconciled per message.id
## Summary

Token totals and costs are inflated roughly **2–4x** for Claude Code sessions: both ingestion paths sum a message's `usage` once per JSONL **record** instead of once per **message.id**. On my dashboard this turned a ~$16k API-equivalent history into a displayed **$32k**.

There is also a second, subtler bug in the opposite direction, and a repair problem for existing databases — details below. I've implemented and tested a fix on a fork (branch linked at the bottom); happy to open a PR if you'd like it.

## Root cause 1 — usage summed per record (inflation, 2–4x)

Claude Code writes one JSONL record per content block, so a single API response (one `message.id`) appears as several records — each carrying a copy of the message's `usage`. Both accumulation sites count every record:

- `server/lib/transcript-cache.js` → `_ingestEntry` (live path)
- `scripts/import-history.js` → `parseSessionFile` and `parseSubagentFile` (import path)

**Reproduction** (any Claude Code session transcript):

```python
import json, sys
from collections import defaultdict
naive, seen, dedup = {}, set(), {}
for line in open(sys.argv[1]):
    try: rec = json.loads(line)
    except: continue
    msg = rec.get("message") or {}
    u = msg.get("usage")
    if not u: continue
    for k in ("input_tokens","output_tokens","cache_read_input_tokens"):
        naive[k] = naive.get(k,0) + u.get(k,0)
    if msg.get("id") in seen: continue
    seen.add(msg.get("id"))
    for k in ("input_tokens","output_tokens","cache_read_input_tokens"):
        dedup[k] = dedup.get(k,0) + u.get(k,0)
print("naive:", naive); print("deduped:", dedup)
```

On a live session of mine, the dashboard's numbers matched the naive sum **to the token** (`cache_read` 18,967,634 naive vs 7,573,797 deduped — 2.5x). 104 of 182 usage records in that file were duplicates.

## Root cause 2 — the duplicates are not identical (undercount if you dedup naively)

First-record-wins dedup is also wrong: streaming writes **partial usage first**, then the final record. Example from a subagent transcript: same `message.id` with `output_tokens: 5` on the first record and `742` on the last. A census across 5,297 local JSONLs found **15,400 message ids whose first and final `output_tokens` differ** (34.7M summed first-record vs 60.8M final-record output; the first record was lower in every case). The **last record per message.id is authoritative**.

## Root cause 3 — existing databases can't self-heal

`replaceTokenUsage` in `server/db.js` is a deliberate high-water mark: when a corrected (lower) re-read arrives, the difference is folded into the `baseline_*` columns, so `effective = current + baseline` never decreases. That's the right call for append-only transcripts, but it means the parser fix alone can never repair historical inflation — the over-count just migrates into the baselines. A repair path that re-derives totals and **zeroes the baselines** is needed, and it must also clear stale bucket rows whose (model/speed/geo/tier) key no longer occurs in the corrected derivation.

## Fix on my fork

Branch: https://github.com/msshives-gif/Claude-Code-Agent-Monitor/tree/fix/dedup-usage-by-message-id

- Last-record-wins reconciliation (`message.id` → counted contribution; a later record retracts and replaces the earlier one) in `TranscriptCache` and both import parsers, with the map seeded across incremental read boundaries via the cached result, tail-capped with `MAX_ARRAY_LEN`.
- `subtractBucket` helper in `server/lib/token-usage.js`.
- Repair path: `node scripts/import-history.js --reconcile-tokens --all --reset-baselines` re-derives every session with a transcript on disk, clears its non-workflow token rows (workflow rows come from run journals, not transcripts, so they're preserved), and writes fresh totals with zeroed baselines (new `resetTokenUsage` statement). Empty corrected derivations still clear stale rows. The repair refuses to run while a dashboard answers on a known port (transcript parsing happens outside the write transaction, so live hook ingestion would race it), and refuses `--dry-run` rather than ignoring it.
- Regression tests in `server/__tests__/usage-dedup.test.js` (identical duplicates, evolving usage, evolution straddling an incremental boundary, id-less records, per-record thinking-block counting preserved).

## Known residual limitations (worth deciding upstream)

- **Live-path key drift under the high-water fold.** If a message's partial record lands in one pricing bucket and its final record in another (tier/speed drift), the corrected parser nets the first bucket to zero — but a live hook that already persisted the partial gets it folded into `baseline_*` by `replaceTokenUsage`, preserving a small over-count (bounded by one partial record per drifted message). A real fix needs the live write path to distinguish "corrected to zero" from "transcript rewritten", which is a design call for you. The periodic reset repair cleans up whatever accumulates.
- Sessions whose project JSONL has been deleted are skipped by the repair (even when a dashboard snapshot exists), so their historical inflation survives.

Full `npm run test:server` passes (952 tests). After the fix + repair, per-session dashboard numbers match an independent per-`message.id` computation exactly; my grand total moved from $32.1k to ~$16k.

Verified against Claude Code transcripts as of 2026-08-15. Thanks for the project — the transcript-cache engineering is genuinely impressive, which is what made the byte-exact match with the naive sum easy to pin down.

### PR DESCRIPTION
**Pull request by hoangsonww**
## Summary

Token totals and costs were inflated **2–4×** because both ingestion paths summed a message's `usage` once per JSONL **record** instead of once per **`message.id`** — Claude Code writes one record per content block and each copies the message's usage. This reconciles usage per message (last record wins) in the live `TranscriptCache` and both import parsers, and adds a repair sweep that heals databases already inflated by the old behaviour.

Reported and diagnosed by @msshives-gif in #293, with a reference implementation on their fork that this PR builds on.

## Changes

- **Reconcile, don't dedup.** The duplicate records are *not* identical — streaming writes partial usage first (`output_tokens: 5`, then the final `742`), so first-record-wins would *undercount*. The **last record for a `message.id` is authoritative**. Each parser keeps a `message.id → counted contribution` map; a later record retracts the earlier contribution via the new `subtractBucket` helper and counts its own.
  - `server/lib/transcript-cache.js` (live path) seeds the map from the cached result across incremental read boundaries, so a message whose records straddle a read still reconciles. The retraction can leave a bucket transiently negative in the incremental chunk; `_merge` nets it out against the cached totals.
  - `scripts/import-history.js` (`parseSessionFile`, `parseSubagentFile`) runs the same reconciliation over a single full pass.
- **Bounded reconciliation window.** New shared `rememberUsageContribution` helper in `server/lib/token-usage.js` keeps the map a tail window in *every* path (default 1000; the cache path uses `TRANSCRIPT_CACHE_MAX_ARRAY_LEN`). A message's records are written adjacently, so a tail window reconciles every real case without letting the map grow with a multi-hundred-MiB transcript — this keeps the import parsers consistent with the cache's existing memory guarantees.
- **Unchanged on purpose:** records with no `message.id` are still counted unconditionally, and per-record content-block processing (thinking counts, `tool_use` extraction) is untouched — each record carries a *distinct* block even when the usage copy is repeated.
- **The repair now runs automatically.** Shipping the parser fix with a manual CLI step would have left almost every existing user on wrong costs indefinitely — `replaceTokenUsage` is a high-water mark, so a corrected (lower) re-read just migrates the over-count into `baseline_*` and the effective number never drops. `repairInflatedTokenTotals()` runs the sweep once per database from `startBackgroundServices`, gated by a `.token-repair-v1.done` marker written only after a completed pass (a crash retries), deferred off the boot path, skipped without consuming the marker while `peersSharingDataDir()` sees another instance, and disabled by `DASHBOARD_TOKEN_REPAIR=0`. It snapshots the table to `token_usage_pre_repair` first, so the pre-repair numbers stay recoverable with plain SQL. A hook landing mid-repair can lose one write, but that self-heals: with baselines zeroed, the next event for that session re-parses the whole transcript and writes the true total.

  Verified on a real 604 MB production database: **66 sessions corrected, 29.92B → 26.41B tokens**, baselines 42.4M → 0, snapshot intact.

- **Repair path for existing databases.** `replaceTokenUsage` is a deliberate high-water mark, so the parser fix alone can never heal historical rows — a corrected (lower) re-read just migrates the over-count into `baseline_*`. New `resetTokenUsage` statement + sweep:

  ```bash
  npm run repair-tokens
  # = node scripts/import-history.js --reconcile-tokens --all --reset-baselines
  ```

  - `--all` widens the sweep from `imported`-marked sessions to every **Claude** session with a transcript on disk (live-ingested rows carry no marker). Transcripts are located by scanning `~/.claude/projects/`, falling back to the `transcript_path` persisted on the session row — so a session imported from a custom directory is covered too.
  - **Codex sessions are excluded** (`provider = 'claude'`). Their usage comes from rollout journals via `upsertCodexTokenDelta`, never from a Claude transcript, so the sweep must neither rebuild nor clear their rows.
  - `--reset-baselines` clears each session's non-workflow `token_usage` rows before writing, so stale buckets whose `(model, speed, geo, tier)` key no longer occurs cannot survive; an empty corrected derivation still clears. Workflow rows (`service_tier = 'workflow'`) come from run journals rather than transcripts and are **preserved**.
  - Refuses to run while a dashboard answers on a known port (transcripts are parsed *outside* the write transaction, so live hook ingestion would race the repair) and rejects `--dry-run` rather than silently ignoring it.
  - The `--reconcile-tokens` banner now names the actual scope/mode instead of always claiming "already-imported sessions".
- **Docs:** new "Per-message usage reconciliation" section in `ARCHITECTURE.md` (including the repair procedure and its two deliberate limitations), module rows in `ARCHITECTURE.md` / `server/README.md`, command tables in `README.md` + VN/CN/KO/ES, `SETUP.md` (CLI + scripts reference), and `wiki/index.html` with `zh`/`vi`/`ko`/`es` translations plus the service-worker cache bump.
- **Release:** patch bump to **v2.0.9** (backward-compatible data-correctness fix plus its remediation script — no new product surface). Synchronized across every version surface; see the audit below.

### Known limitations (deliberate, documented)

- Sessions whose transcript has been **deleted** (or whose stored path is stale) are skipped and counted in `missingFiles` — their totals cannot be re-derived, and dropping the rows would destroy real usage.
- If a message's partial and final records ever land in **different pricing buckets**, the live writer's high-water fold preserves the partial (bounded by one record per drifted message). Re-running the repair clears whatever accumulates.

### v2.0.9 release audit

Every surface verified against the release checklist with fresh reads:

| Surface | State |
|---|---|
| Root `package.json`, lockfile `.version` + `packages[""]` | 2.0.9 |
| `desktop/package.json`, lockfile `.version` + `packages[""]` | 2.0.9 |
| Live OpenAPI spec + generated `openapi.yaml` (`info.version`, `HealthResponse` example) | 2.0.9 |
| 14 Claude + 14 Codex plugin manifests (`extensions:sync`) | 2.0.9 |
| `docker-compose.yml` (dashboard + MCP images) | 2.0.9 |
| Helm `Chart.yaml` `version` + `appVersion` | 2.0.9 |
| Kubernetes labels, images, kustomize `newTag` (18 files) | 2.0.9 |
| `deploy.sh` `--tag`/substitution, `DEPLOYMENT.md`, `docs/DEPLOYMENT.md` | 2.0.9 |
| Version-sensitive UI snapshot | 2.0.9 |
| `ccam version` | `ccam 2.0.9` |
| `client` 1.0.0 · `mcp` 1.0.0 · `monitoring` 1.0.0 · `vscode-extension` 1.2.4 | intentionally unchanged |

- Helm per-env values pin no tag, so `_helpers.tpl`'s `default .Chart.AppVersion` resolves both images to 2.0.9.
- `.claude-plugin/marketplace.json` carries no version field — nothing to bump.
- Re-running `extensions:sync` and `openapi:yaml` produces **no diff**: the generated artifacts are at their fixed point.
- `client/dist` is gitignored and Vite injects `__APP_VERSION__` from the root manifest at build time, so `npm run build` was run to confirm the bundle emits 2.0.9 (Settings compares it against the server version to warn about a stale UI build).
- No git tag or GitHub release was created.

**The audit found a hole in the guard itself**, now closed: nothing verified the plugin manifests, so a bump that skipped `extensions:sync` would leave 28 manifests advertising the previous release with no test failing. `release-version-consistency.test.js` gains three checks — plugin manifests carry the root version, the marketplace catalog lists every bundled plugin, and the independently shipped packages stay *off* the root version. Each was mutation-tested (reverting one manifest to 2.0.8 fails the first; setting `client` to 2.0.9 fails the third), so none pass vacuously.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (no functional changes)
- [x] Documentation update
- [ ] Infrastructure / CI / DevOps
- [ ] Dependency update

## How to Test

1. `npm run test:server` — **1016 tests, 1015 pass / 0 fail / 1 skipped** (up from 967). New coverage:
   - `server/__tests__/usage-dedup.test.js` (live cache): identical duplicates, evolving usage, evolution straddling an incremental read boundary, id-less records, per-record thinking counts preserved.
   - `server/__tests__/usage-dedup-import.test.js` (import + repair + cost): both parsers reconcile last-record-wins, the bounded window evicts oldest and refreshes recency, `resetTokenUsage` zeroes baselines where `replaceTokenUsage` folds them, the repair sweep end-to-end (inflated rows corrected, stale bucket rows cleared, **workflow rows preserved**, empty derivations still cleared, missing transcripts skipped), the corrected totals price strictly below the naive sum, a DB-priced session matches one priced straight from the parsed transcript, and `combineSessionTokens` adds already-reconciled parent + subagent buckets without re-counting.
   - `server/__tests__/token-usage-helpers.test.js` (new, 32 tests): the shared helper module had no direct unit coverage despite both ingestion paths depending on it. Normalizers (including `not_available` geo and non-rate-changing `priority` tier), bucket keying, `extractUsageFields` across every legacy shape (explicit total beats breakdown, no-breakdown counts as all-5m, the 1h subset clamped so the 5m portion can't go negative at pricing time), `accumulateBucket`/`subtractBucket` as an **exact inverse pair** that deliberately does not clamp at zero, and `rememberUsageContribution` eviction/recency/by-reference semantics.
2. `npm run test:client` — 330 tests pass (36 files), including the wiki i18n coverage/tag/asset-version test and the regenerated screen snapshot.
3. `npm run build`, `npm run mcp:typecheck`, `npm run format:check`, `npm run extensions:validate`, and `bash .claude/skills/file-headers/scripts/check-headers.sh` — all clean.
4. Repair the real database (destructive to `token_usage`; **stop the dashboard first**):
   ```bash
   npm run repair-tokens
   ```
   Then compare a session's dashboard total against an independent per-`message.id` computation over its JSONL — they should match exactly, and the grand total should drop to roughly half of the pre-fix figure.

### Verification against real data

Beyond the unit suite, the reconciliation was cross-checked against the real transcript corpus on the author's machine using an independent reference implementation that shares no code with `server/lib/token-usage.js`:

- **295 transcripts, 135 carrying usage, 24,958 usage records.** 13,125 of those records (**52.6%**) were duplicate copies of a message's usage, and **1,010 of the duplicates carried different usage** than the earlier record — confirming last-record-wins is required and that first-record-wins would undercount.
- Naive per-record sum: **8.71B** input+output+cache_read tokens vs a correct **4.53B** — a **1.92×** inflation on this corpus.
- The live cache (full read), both import parsers, and a 60-file incremental-read replay **all match the reference exactly on every file**, with no negative bucket anywhere.
- **Window assumption measured:** across all 13,125 duplicate records the span between a message's records was always exactly **1** distinct `message.id` — they are strictly adjacent. The 1000-entry window carries a ~1000× margin; `USAGE_RECONCILE_WINDOW` records that measurement.
- **Full-stack check** (real transcripts → actual import pipeline → `token_usage` rows → actual `calculateCost`): across 25 real sessions, stored tokens matched the independent computation exactly, per-session cost matched independently re-derived rate math (fast / US-residency / batch modifiers, 5m-vs-1h cache split, web-search surcharge, code-execution allowance), the grand total equalled both the unrounded per-session sum and the independent total, no negative row reached the database, the repair was idempotent, a plain reconcile could **not** heal simulated inflation (the high-water mark holds, as designed), and the reset repair restored the exact correct cost.

### Review follow-ups (CodeRabbit)

Seven findings across two review rounds — **all addressed, all threads resolved.** Round one in `75582c0`, round two in `b727210`.

1. **Repair scope overstated** *(Major)* — real gap, not just wording. The sweep only matched `<proj>/<sid>.jsonl`, skipping Claude sessions whose transcript lives elsewhere. Now falls back to `sessions.transcript_path`. **Widening it exposed a latent hazard**: Codex sessions also carry a `transcript_path`, so a naive fallback would have fed them to `parseSessionFile` and — since reset mode clears rows before writing — **deleted real Codex usage**. Added an explicit `provider = 'claude'` filter. Measured locally: repair coverage 186 → **191** Claude sessions, while the filter blocks **689** Codex sessions the fallback would otherwise have swept in, **533** of which hold `token_usage` rows.
2. **Atomic repair lock** *(Major)* — partially addressed, deliberately deferred. Fixed the real defect: the probe treated *any* failure as "no server", so a timeout against a busy dashboard read as safe. Now only `ECONNREFUSED` proves nothing is listening; timeouts, other transport errors, and non-2xx all fail closed. The full cross-process lock would have to be checked before every live token write — the hot ingestion path — which is an architectural decision with its own failure modes (stale locks after a crash, ordering against the SQLite transaction) and belongs in its own change, not bundled into a token-accounting fix.
3. **Shared window constant** *(Minor)* — test half applied, production half declined; reviewer agreed the finding "does not apply". Window tests now derive fixture sizes from the effective window (plus a new inside-the-boundary case). The live cache keeps sizing from `TRANSCRIPT_CACHE_MAX_ARRAY_LEN` because switching to the fixed constant would silently remove that documented env override.
4. **Wiki repair-scope wording** *(Minor)* — applied, extended to name the Claude-only scope; `zh`/`vi`/`ko`/`es` rekeyed and the service-worker cache bumped.
5. **Release guard gaps** *(Minor)* — both plugin formats now **required** per plugin rather than probed, and both `.claude-plugin` *and* `.agents/plugins` marketplace catalogs are validated against the plugin set.

Round two, on the tests added in round one:

6. **Repair aborted when the scan root was absent** *(Major)* — a defect the persisted-path fallback itself introduced. `reconcileTokens` returned early when `~/.claude/projects/` did not exist, which was correct while the directory scan was the *only* transcript source; with `transcript_path` as a second source it silently stranded every session whose transcript lives outside that tree (fresh machine, non-default `CLAUDE_HOME`, or a container holding only the DB plus mounted transcripts). A missing scan root is now an empty listing and the sweep continues.
7. **Window test didn't test the boundary** *(Minor)* — the filler was sized as a fraction of the window, which stops probing the edge as the window grows and **fails outright at `TRANSCRIPT_CACHE_MAX_ARRAY_LEN=1`** (one filler evicts the tracked message, double-counting its final record). Now `WINDOW - 1`: the largest separation that stays retractable, correct at any window size.

Every new test is mutation-verified — removing the `transcript_path` fallback, the `provider` filter, or the missing-root handling each fails its own test, and the previous fraction-based filler fails at `WINDOW=1` where the new one passes. All three reconciliation suites were additionally re-run end to end under `TRANSCRIPT_CACHE_MAX_ARRAY_LEN=1` (65 tests) to confirm nothing else depends on the default window size.

## Checklist

- [x] I have read the [contributing guidelines](https://github.com/hoangsonww/Claude-Code-Agent-Monitor/blob/master/.github/CONTRIBUTING.md)
- [x] I have signed the [CLA](https://github.com/hoangsonww/Claude-Code-Agent-Monitor/blob/master/CLA.md)
- [x] My code follows the project's coding standards
- [x] I have added/updated tests that prove my fix or feature works
- [x] All new and existing tests pass (`npm test`)
- [x] Code is formatted (`npm run format:check`)
- [x] I have updated documentation where necessary

Closes #293

### CODE REVIEW

File: ARCHITECTURE.md [hunk 1]
```diff
@@ -1322,6 +1322,72 @@ writer stores main-only tokens; `importSession` stores main+subagents)
...
+baselines via the `resetTokenUsage` statement:
+
+```bash
+npm run repair-tokens
+# = node scripts/import-history.js --reconcile-tokens --all --reset-baselines
+```
+
+- `--all` widens the sweep from `imported`-marked sessions to every session
+  whose transcript is still on disk (live-ingested rows carry no marker).
+- `--reset-baselines` clears each session's non-workflow `token_usage` rows
```
[the remark this answers is not in the document]
Comment (hoangsonww) [reply ARCHITECTURE.md]: Fixed in 75582c0 — and this turned out to be more than a docs issue, thanks for catching it.

You're right that the scan only matched `<proj>/<sid>.jsonl`. I took the "resolve persisted transcript paths" option rather than weakening the docs, so `--all` now falls back to `sessions.transcript_path` (stale paths count as missing).

**But implementing that naively would have destroyed data**, which is worth recording here. Codex sessions *do* carry a `transcript_path`, yet their `token_usage` comes from rollout journals via `upsertCodexTokenDelta` — never from a Claude JSONL transcript. Resolving stored paths would have fed them to `parseSessionFile`, and since reset mode clears a session's non-workflow rows before writing, it would have deleted real Codex usage and replaced it with a bogus derivation. They were only excluded before *by accident*, because their transcripts sit outside the scanned tree. The sweep now filters `provider = 'claude'` explicitly.

Measured on my local database:

| | sessions |
|---|---|
| Claude repaired before (scan only) | 186 |
| Claude repaired after (+ `transcript_path`) | **191** |
| Codex the fallback would have swept in | **689** — blocked by the provider filter |
| …of those, holding `token_usage` rows | **533** — would have been cleared |

Both behaviours are mutation-tested: the fallback test fails without the fallback, and the Codex test fails without the provider filter.

Docs corrected to state the scope precisely (Claude-only, non-workflow rows only, both lookup paths, and the `missingFiles` case) across `ARCHITECTURE.md`, `SETUP.md`, `server/README.md`, `README.md` + VN/CN/KO/ES, and the wiki.

File: scripts/import-history.js [hunk 2]
```diff
@@ -2127,6 +2192,35 @@ if (require.main === module) {
...
+                `A dashboard server is running on port ${port} — stop it before running` +
+                  ` --reconcile-tokens --reset-baselines (live hook ingestion races the repair).`
+              );
+              process.exit(1);
+            }
+          } catch {
+            /* no server answering on this port — safe */
+          }
+        }
+      }
```
[the remark this answers is not in the document]
Comment (hoangsonww) [reply scripts/import-history.js:2283]: Partially addressed in 75582c0; declining the full lock for now.

You identified a real defect in the guard, which I've fixed: it treated **any** probe failure as "no server", so a timeout against a busy-but-alive dashboard read as safe. Only `ECONNREFUSED` now counts as proof nothing is listening — a timeout or any other transport error fails closed, and a non-2xx response also blocks (a degraded dashboard is still a writer). Timeout raised to 2s.

I'm not implementing the shared lock in this PR. It would require the dashboard to check a lock at startup *and* before every live token write — that's the hot ingestion path, and a new cross-process lock protocol there is an architectural decision with its own failure modes (stale locks after a crash, ordering with the SQLite transaction) that belongs in its own change with the maintainer's input, not bundled into a token-accounting bug fix.

The residual window is: probe passes → a dashboard starts → repair completes. For a one-time, manually-run repair that already refuses on any ambiguous probe result and is documented to require stopping the dashboard, I think that's an acceptable trade rather than something to fix by expanding this PR's blast radius. Happy to file it as a follow-up issue if you'd like it tracked.

File: server/lib/transcript-cache.js [hunk 3]
```diff
@@ -550,14 +570,44 @@ class TranscriptCache {
...
+      // Bounded tail window (see rememberUsageContribution) — a message's
+      // records sit adjacent, so a tail window reconciles every real case
+      // without letting the map grow with the transcript.
+      rememberUsageContribution(
+        state.usageByMessageId,
+        usageMsgId,
+        { key, model, speed, geo, tier, fields },
+        MAX_ARRAY_LEN
+      );
+    }
```
[the remark this answers is not in the document]
Comment (hoangsonww) [reply server/lib/transcript-cache.js:610]: Test half applied in 75582c0; declining the production half.

The window tests no longer hard-code `1100`/`1000` — they derive fixture sizes from the effective window, and I added a complement case just inside the window so both sides of the boundary are pinned.

I did **not** switch the live cache to `USAGE_RECONCILE_WINDOW`, because that would be a behaviour regression. The cache deliberately sizes this from `TRANSCRIPT_CACHE_MAX_ARRAY_LEN`, which is an existing documented env override users can tune; passing the fixed shared constant instead would silently remove that. The test now reads the same env var, so production and test stay aligned either way.

File: wiki/index.html [hunk 4]
```diff
@@ -2097,6 +2097,16 @@ <h2>Scripts Reference</h2>
...
                 </tr>
+                <tr>
+                  <td><code>reconcile-tokens</code></td>
+                  <td><code>npm run reconcile-tokens</code></td>
+                  <td>Refresh token totals for imported sessions from their transcripts. Never lowers an existing total</td>
+                </tr>
+                <tr>
+                  <td><code>repair-tokens</code></td>
+                  <td><code>npm run repair-tokens</code></td>
+                  <td>Re-derive token totals for every session whose transcript is still on disk and zero the compaction baselines. One-time repair for databases inflated by the old per-record usage sum; stop the dashboard before running it</td>
```
[the remark this answers is not in the document]
Comment (hoangsonww) [reply wiki/index.html]: Applied in 75582c0, with one addition to your wording.

Took the "non-workflow … preserve workflow rows" precision, and also scoped it to **Claude** sessions — the same commit excludes Codex sessions from the sweep (their usage comes from rollout journals, not transcripts, so the repair must neither rebuild nor clear their rows). Final text:

> Re-derive non-workflow token totals for Claude sessions whose transcripts are still on disk and reset the compaction baselines, preserving workflow and Codex rows. One-time repair for databases inflated by the old per-record usage sum; stop the dashboard before running it

The `zh`/`vi`/`ko`/`es` entries were rekeyed to the new English string and the service-worker cache bumped (`wiki-v100`, `i18n-content.js?v=81`), so returning visitors get it. `client/tests/wiki-i18n.test.ts` passes.

File: server/__tests__/release-version-consistency.test.js [hunk 5]
```diff
@@ -89,6 +89,65 @@ describe("release version consistency", () => {
...
+    const manifests = [];
+    for (const entry of fs.readdirSync(pluginsDir, { withFileTypes: true })) {
+      if (!entry.isDirectory()) continue;
+      for (const format of [".claude-plugin", ".codex-plugin"]) {
+        const manifest = path.join("plugins", entry.name, format, "plugin.json");
+        if (fs.existsSync(path.join(ROOT, manifest))) manifests.push(manifest);
+      }
+    }
+
+    assert.ok(manifests.length > 0, "plugin manifests must exist");
```
[the remark this answers is not in the document]
Comment (hoangsonww) [reply server/__tests__/release-version-consistency.test.js]: Both applied in 75582c0.

- The manifest loop no longer probes with `existsSync` — every bundled plugin must ship **both** `.claude-plugin` and `.codex-plugin` manifests, asserted before the version check, so a release that drops a Codex manifest now fails.
- The catalog test iterates **both** `.claude-plugin/marketplace.json` and `.agents/plugins/marketplace.json`, asserting the identical plugin set against the plugin directories.

Both were mutation-tested when added (reverting one manifest to `2.0.8` fails the version check; setting `client` to the root version fails the independent-subproject check), so none of these pass vacuously.

File: scripts/import-history.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (hoangsonww) [reply scripts/import-history.js:2283]: No follow-up issue for now, thanks — we'll leave the residual probe→start race documented here and in `ARCHITECTURE.md` rather than tracked as open work.

The reasoning: `repair-tokens` is a one-time, manually-run maintenance command that now fails closed on every ambiguous probe result and is documented to require stopping the dashboard first. The remaining window needs someone to start a dashboard *during* the repair. That doesn't justify putting a new cross-process lock check on the live token-write path, which is where the cost of the full design would land.

Leaving this thread unresolved is fine — it's an accurate record of a known, accepted limitation.

File: server/__tests__/usage-dedup-import.test.js [hunk 6]
```diff
@@ -0,0 +1,565 @@
...
+  });
+
+  it("repairs a session reachable only through its persisted transcript_path", async () => {
+    await reconcileTokens(dbModule, { all: true, resetBaselines: true });
+    const row = rawRow(OUTSIDE, MODEL);
+    assert.ok(row, "the session must have been re-derived, not skipped");
+    assert.equal(row.output_tokens, EXPECTED.output);
+    assert.equal(row.cache_read_tokens, EXPECTED.cacheRead);
+    assert.equal(row.baseline_output, 0);
+  });
```
[the remark this answers is not in the document]
Comment (hoangsonww) [reply server/__tests__/usage-dedup-import.test.js:534]: Good catch — fixed in b727210.

You're right, and it's a defect the persisted-path fallback introduced. `if (!fs.existsSync(PROJECTS_DIR)) return counters;` was correct while the directory scan was the *only* way to find a transcript; once `transcript_path` became a second source, bailing there silently stranded every session whose transcript lives outside that tree — a fresh machine, a non-default `CLAUDE_HOME`, or a container holding only the database plus externally mounted transcripts. A missing scan root is now an empty directory listing and the provider-filtered persisted-path sweep continues.

Added the missing-root condition as its own test (moves the whole projects tree aside for the duration of the sweep, restores it in a `finally`). Mutation-verified: restoring the early return fails it.

File: server/__tests__/usage-dedup.test.js [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+    const filler = Math.max(1, Math.floor(WINDOW / 2));
+    const lines = [assistantRecord("msg_near", near)];
+    for (let i = 0; i < filler; i++) lines.push(assistantRecord(`inside_${i}`, tiny));
+    lines.push(assistantRecord("msg_near", near));
+    fs.writeFileSync(p, lines.join("\n") + "\n");
+
+    const result = new TranscriptCache().extract(p);
+    assert.equal(sumField(result, "output"), near.output_tokens + filler);
+    assert.deepEqual(negativeBuckets(result), []);
+  });
```
[the remark this answers is not in the document]
Comment (hoangsonww) [reply server/__tests__/usage-dedup.test.js:349]: Applied in b727210 — your analysis was exactly right.

I confirmed the failure before fixing it: at `TRANSCRIPT_CACHE_MAX_ARRAY_LEN=1` the single filler evicts `msg_near`, so its final record is counted twice and the assertion fails. `WINDOW - 1` is the correct boundary — the map then holds exactly `WINDOW` entries (the tracked message plus the fillers) and evicts nothing, so it pins the largest separation that remains retractable at any window size, including 1.

Also re-ran all three reconciliation suites end to end under `TRANSCRIPT_CACHE_MAX_ARRAY_LEN=1` (65 tests) to confirm nothing else silently depended on the default window.

On the ast-grep `detect-non-literal-fs-filename` warning: not applicable here — `p` is `path.join(tmpDir, "<literal>")` where `tmpDir` comes from `fs.mkdtempSync`, with no external input anywhere in the chain.

File: server/__tests__/usage-dedup-import.test.js [hunk 8]
```diff
@@ -0,0 +1,604 @@
...
+    // Move the whole projects tree aside for the duration of this sweep.
+    const projectsRoot = path.join(CLAUDE_HOME, "projects");
+    const stashed = path.join(TMP_ROOT, "projects-stashed");
+    fs.renameSync(projectsRoot, stashed);
+    assert.equal(fs.existsSync(projectsRoot), false, "scan root must be absent for this test");
+
+    try {
+      await reconcileTokens(dbModule, { all: true, resetBaselines: true });
+    } finally {
+      fs.renameSync(stashed, projectsRoot);
```
[the remark this answers is not in the document]
Comment (hoangsonww) [reply server/__tests__/usage-dedup-import.test.js]: Fixed in 7ad6fd3. The `try` now starts before the rename and tracks whether the move actually happened, restoring only in that case — so a failure of the assertion between the two can no longer leave `CLAUDE_HOME/projects` stashed and break every later test in the file.
